File: app/src/main/java/com/project/poupay/view/ListContentAdapter.java
package com.project.poupay.view;

import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.project.poupay.R;

import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class ListContentAdapter extends RecyclerView.Adapter<ListContentAdapter.ViewHolder> {

    private final List<ContentReportItem> mItems;

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private final ImageView viewDrawable;
        private final TextView viewTitle, viewValue;

        public ViewHolder(View view) {
            super(view);
            viewDrawable = view.findViewById(R.id.ContentReportItem_Image);
            viewTitle = view.findViewById(R.id.ContentReportItem_Title);
            viewValue = view.findViewById(R.id.ContentReportItem_Value);
        }

        public void setup(ContentReportItem item) {
            double total = 0;
            viewDrawable.setImageResource(item.getImageID());
            viewTitle.setText(item.getTitle());
            viewValue.setText(NumberFormat.getCurrencyInstance(Locale.getDefault()).format(item.getValue()));
            viewValue.setTextColor(item.getValue() < 0 ? Color.parseColor("#E57373") : Color.parseColor("#0FA958"));
        }
    }

    public ListContentAdapter() {
        mItems = new ArrayList<>();
    }

    public void add(ContentReportItem newItem) {
        mItems.add(newItem);
        notifyItemInserted(getItemCount() - 1);
    }

    public void add(List<ContentReportItem> newItems) {
        for (ContentReportItem item : newItems) add(item);
    }

    public void clear() {
        int count = getItemCount();
        mItems.clear();
        notifyItemRangeRemoved(0, count);
    }

    public Double getValueByContentTitle(String title) {
        Double value = null;
        for (ContentReportItem item : mItems) if (item.getTitle().equals(title)) return item.getValue();
        return value;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.view_contentreportitem, viewGroup, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, final int position) {
        viewHolder.setup(mItems.get(position));
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }
}

File: app/src/main/java/com/projeto/poupay/view/ListContentAdapter.kt
package com.projeto.poupay.view

import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import com.projeto.poupay.R
import java.text.NumberFormat
import java.util.Locale

class ListContentAdapter : RecyclerView.Adapter<ListContentAdapter.ViewHolder>() {

    private var mItens: MutableList<ContentReportItem> = mutableListOf()
    private var onItemClickListener: (item: ContentReportItem) -> Unit = {}

    class ViewHolder(val view: View) : RecyclerView.ViewHolder(view) {
        private val viewRootLayout: ConstraintLayout
        private val viewDrawable: ImageView
        private val viewTitle: TextView
        private val viewValue: TextView
        private var onItemClickListener: (item: ContentReportItem) -> Unit = {}

        init {
            with(view) {
                viewRootLayout = findViewById(R.id.ContentReportItem_Root)
                viewDrawable = findViewById(R.id.ContentReportItem_Image)
                viewTitle = findViewById(R.id.ContentReportItem_Title)
                viewValue = findViewById(R.id.ContentReportItem_Value)
            }
        }

        fun setup(item: ContentReportItem) {
            viewRootLayout.setOnClickListener { onItemClickListener.invoke(item) }
            viewDrawable.setImageResource(item.getImageID())
            viewTitle.text = item.subtitule.ifEmpty { item.title }
            viewValue.text = NumberFormat.getCurrencyInstance(Locale.getDefault()).format(item.value)
            viewValue.setTextColor(if (item.value < 0) Color.parseColor("#E57373") else Color.parseColor("#0FA958"))
        }

        fun setOnItemClickListener(listener: (item: ContentReportItem) -> Unit) {
            this.onItemClickListener = listener
        }

    }

    fun setOnItemClickListener(listener: (item: ContentReportItem) -> Unit) {
        this.onItemClickListener = listener
    }

    fun add(newItem: ContentReportItem) {
        mItens.add(newItem)
        notifyItemInserted(itemCount - 1)
    }

    fun add(newItems: List<ContentReportItem>) {
        for (item in newItems) {
            add(item)
        }
    }

    fun getByTitle(): List<ContentReportItem> {
        val filteredList = mutableListOf<ContentReportItem>()
        mItens.forEach { superItem ->
            var alreadyHasItem = false
            filteredList.forEach { item ->
                if (superItem.title == item.title) {
                    alreadyHasItem = true
                    item.value += superItem.value
                }
            }
            if(!alreadyHasItem){
                filteredList.add(superItem.copy())
            }
        }
        return filteredList
    }

    fun clear() {
        val count = itemCount
        mItens.clear()
        notifyItemRangeRemoved(0, count)
    }

    fun getValueByContentTitle(title: String): Double? {
        val internalItems = getByTitle()
        for (item in internalItems) {
            if (item.title == title) {
                return item.value
            }
        }
        return null
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.view_contentreportitem, parent, false)
        return ViewHolder(view)
    }

    override fun getItemCount(): Int {
        return mItens.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setup(mItens[position])
        holder.setOnItemClickListener(this.onItemClickListener)

    }
}